### PR TITLE
Virion v3.0

### PR DESCRIPTION
**Pull request by SOF3**
[Rendered](https://github.com/poggit/support/blob/virion-3.0/virion.md)

### CODE REVIEW

**Review by dktapps — COMMENTED** [review]
Looks interesting, I've only briefly skimmed over it but I do like the usage of Composer for this.

Perhaps we can also enable non-virion libraries to be shaded this way by having some kind of property override. Some libraries will be able to be shaded even if they don't declare it.

**Comment by SOF3** [comment]
For including non virion libraries, I am not sure how we would distinguish between actual libraries and plugins provided in deps. Maybe composer.json type=library of the dep?

**Comment by dktapps** [comment]
I think there are ways to set/override package metadata from the parent composer.json.

**Review by NhanAZ — APPROVED** [review]

**Review by Ifera — APPROVED** [review]

File: virion.md [hunk 1]
```diff
@@ -1,161 +1,171 @@
...
+so different libraries from the same plugin can interact with the same interface
+but cannot interact with libraries included from other plugins.
 
-## Appendix
-### Syntactic references
-Basically, a syntactic reference is defined as:
+## Virion user guide
 
-> A `T_NAMESPACE` (`namespace`), `T_USE` (`use`) or `T_NS_SEPARATOR` (`\`) token, or `T_USE` + `T_FUNCTION` (`use function`) or `T_USE` + `T_CONST` (`use const`), followed by a consecutive series of `T_STRING` (a non-keyword set of characters, i.e. part of the namespace) and `T_NS_SEPARATOR` (`\`).
+To use a virion, firstt create a composer.json for your project:
```
Comment (jasonwynn10) [inline virion.md:111]: small typo

File: virion.md [hunk 2]
```diff
@@ -1,161 +1,171 @@
...
-\name\space\Clazz::method();
+For each virion, add an entry to the `require` attribute
+just like normal composer libraries:
+
+```diff
+  {
+    "name": "author/project",
+    "require": {
++     "sof3/await-generator": "^3.0.0",
+      "pmmp/pocketmine-mp": "^4.13.0"
```
Comment (jasonwynn10) [inline virion.md:133]: does the virion shader remove this as a dependency or are all of pocketmine's classes and libraries shaded into the plugin?

File: virion.md [hunk 3]
```diff
@@ -1,161 +1,171 @@
...
 
-// syntactic references, from the scope of a class method in poggit\libasynql\SqlResult
-$class = new \ReflectionClass(self::class);
-$class = new \ReflectionClass(static::class);
-$class = new \ReflectionClass(get_class());
-$class = new \ReflectionClass(__CLASS__);
-$class = new \ReflectionClass(__NAMESPACE__ . "\\SqlResult"); // a boundary case barely acceptable
+Version 1 was largely coupled with Poggit, resulting in vendor lock-in and poor tooling.
+Its dependency declaration was coupled with Poggit
+and was not supported by most extenral tools,
```
Comment (jasonwynn10) [inline virion.md:160]: small typo

File: virion.md [hunk 4]
```diff
@@ -1,161 +1,171 @@
...
 
-Basically, a virion release has the same format as the plugin release, except with four differences:
-- `plugin.yml` becomes the `virion.yml`
-- `resources` are not supported (may be added in the future).
-- A file called `virion.php` and a file called `virion_stub.php` are included into the phar. These files can be [downloaded from the Poggit Git repo](https://github.com/poggit/poggit/tree/beta/assets/php/).
-  - The `virion_stub.php` provides a command-line user interface for virion injection
-  - The `virion.php` contains the code that actually injects the virion.
-- The phar stub should load the `virion.php` directly. This is an example valid stub:
+A virion MAY declare `require` dependencies that also declare the `x-virion` field.
+Non-virion libraries MUST NOT be included
```
Comment (jasonwynn10) [inline virion.md:56]: For clarification, does this mean not included in the plugin's composer manifest or the virion's manifest?

**Comment by JaxkDev** [comment]
Any chance of moving away from the unique naming conventions, eg renaming `antigen` to something like `namespace`

File: virion.md (hunk 2 above)
Comment (SOF3) [reply virion.md:133]: @jasonwynn10 I haven't updated the spec yet, but I plan to utilize the fact that type=project in PM will exclude this dependency.

File: virion.md (hunk 4 above)
Comment (SOF3) [reply virion.md:56]: both, with the exception of platform dependencies like `php` which should be ignored. I will clarify this.

**Comment by dktapps** [comment]
> Any chance of moving away from the unique naming conventions, eg renaming `antigen` to something like `namespace`

+1 for this, I never found these words to be helpful

**Comment by dktapps** [comment]
`unshaded-namespace-prefix`?

**Comment by dktapps** [comment]
Perhaps just `root-namespace` would also be sufficient. Seems descriptive enough to me.

**Comment by SOF3** [comment]
Updated, please have another look.

**Comment by dktapps** [comment]
I made some comments on the individual commits

**Review by dktapps — COMMENTED** [review]
So, are we making `x-virion` mandatory or no?

**Comment by SOF3** [comment]
Currently it is mandatory, but it might be optional in future versions.

**Comment by dktapps** [comment]
I don't think it's necessary to restrict the library type if we're requiring x-virion to be declared.

**Comment by SOF3** [comment]
If we allow non-`x-virion` libraries to be included in the future, wouldn't that become a breaking change?

**Comment by dktapps** [comment]
I was under the impression that the consensus was that it's not practical to allow libraries without `x-virion`, and that users should fork non-virion-supported libraries to add it anyway.

**Comment by SOF3** [comment]
alright, I suppose for now it's reasonable to limit to libraries with `x-virion` and just ignore `type`.

**Comment by SOF3** [comment]
Actually I've been wondering if we should simply require `"type": "virion"`. But we still need the `x-virion` tag because of the requirement to specify custom fields.

**Comment by SOF3** [comment]
Actually the `x-` prefix is not really so standard. What about using the `extra` field instead? https://getcomposer.org/doc/04-schema.md#extra

**Comment by dktapps** [comment]
I think putting it under `extra` is probably a good idea.

Virions should be able to be consumed as libraries if they don't use the PM API, so I don't think type: virion should be required. Virions in this scenario are essentially just libraries with virion shading metadata.

**Pull request merged**